src/Storages/StorageS3.cpp

```diff
@@ -1233,11 +1286,21 @@ void ReadFromStorageS3Step::initializePipeline(QueryPipelineBuilder & pipeline,
     createIterator(nullptr);
 
     size_t estimated_keys_count = iterator_wrapper->estimatedKeysCount();
-    if (estimated_keys_count > 1)
-        num_streams = std::min(num_streams, estimated_keys_count);
-    else
-        /// Disclosed glob iterator can underestimate the amount of keys in some cases. We will keep one stream for this particular case.
-        num_streams = 1;
+    const auto glob_iter = std::dynamic_pointer_cast<StorageS3Source::DisclosedGlobIterator>(iterator_wrapper);
+
+    if (!(glob_iter && glob_iter->hasMore()))
+    {
+        if (estimated_keys_count > 1)
+            num_streams = std::min(num_streams, estimated_keys_count);
+        else
+        {
+            /// The amount of keys (zero) was probably underestimated. We will keep one stream for this particular case.
+            num_streams = 1;
+        }
+    }
+     /// OTHERWISE, 1000 files were listed, but we cannot make any estimation of _how many_ there are (because we list bucket lazily);
+     /// If there are more objects in the bucket, limiting the number of streams is the last thing we may want to do
+     /// as it would lead to serious (up to <max_threads> times) reading performance degradation.
```

Comment:
By "performance degradation" do you mean that files are going to be fetched one by one in sequential order?

Reply:
Yes, exactly